### ISSUE TITLE
Gradle should not fail if it cannot process a .sha1 file during dependency cache access

### ISSUE THREAD

**Issue by big-guy** [gradle/gradle#4893]
@aplatypus commented on [Sun Mar 25 2018](https://github.com/gradle/gradle-native/issues/593)

### Expected Behavior

- Gradle v4.6  `build` should complete the command phrase below without error. 
- Gradle **v4.4** performs `build` correctly.
- The so called _missing_ dependency is present in `lib/` area
- Line **109** (marked below) should complete.

Gradle script:

```
    jar
    {
        manifest
        {
            attributes( "Main-Class": mainClassName )
/* 109 */       attributes( "Class-Path": 'lib/' + ( configurations.runtime.collect { it.getName() }.join(' lib/') ) )
            attributes( "Codebase": "*" )
            attributes( "Permissions": "all-permissions" )
        }
    }

    assemble.dependsOn distDirectory
```

### Current Behavior

- Under Gradle 4.6 that build fails on line **109** 
- The error says Gradle can't resolve dependency: `ha-api-3.1.9.hk2-jar`
-  This JAR is in the lib directory and also successfully used/build in other peer projects
```
FAILURE: Build failed with an exception.

* Where:
Build file 'D:\work\products\FourAbs\FourAbsSrcPlugin\build.gradle' line: 109

* What went wrong:
A problem occurred evaluating project ':FourAbsSrcPlugin'.
> Could not resolve all files for configuration ':FourAbsSrcPlugin:runtime'.
   > Could not download ha-api.hk2-jar (org.glassfish.ha:ha-api:3.1.9)
      > Failed to download SHA1 for resource 'http://devrepo.4tel.int:8081/nexus/content/groups/public/org/glassfish/ha/ha-api/3.1.9/ha-api-3.1.9.hk2-jar'.
         > For input string: "projec"
```

- this is a regression error.  The same script works and build using Gradle v4.4.  I swapped Gradle versions several times to see if there was an  option or something.  Every time; v4.6 failed, v4.4 works.

### Context

-  We will be unable to use v4.6 until this error is fixed.
- For now we are staying with v4.4 which works.

### Steps to Reproduce (for bugs)

- Unable to provide an example at this point, this is one sub-project  in a large build and I we have deadlines.
- The principle problem is a difference between working behaviour in 4.4 and non-working in 4.6

### Your Environment

 * Build scan URL: 
    https://gradle.com/s/zi3kw72iwzbiq

Windows 10, Java 1.8 u162

**Comment by aplatypus** [comment]
No change.  Others had this issue they just ignored it.

**Comment by jjohannes** [comment]
@aplatypus I need some more context to reproduce this. In general the script above works for me with `4.6`. If I add this for example:

```
repositories {  jcenter() }
dependencies {
    compile 'org.glassfish.ha:ha-api:3.1.9'
}
```

But that is without using the `hk2` classifier. I am not sure what `org.glassfish.ha:ha-api:3.1.9:hk2` is. Is that something you are building internally or can I find that on a public repository to test with? Otherwise, maybe there is an issue with your internal repository and how it delivers the SHA1? Or can you reproduce the problem with another module that is on a public repository (so that we can reproduce it as well)?

Side note: Instead of `runtime`, you should use `runtimeClasspath`.

**Comment by aplatypus** [comment]
@jjohannes ... Hi, responses in order:

1. Nexus SHA1, I could look again.  Though as I said no one would change it.  The Nexus repository is just set-up for pass through cache from `maven central` or `jcenttre`.

2.   This JAR is not a directly specified dependency, it is downstream from a WSDL2Java plugin we are using for WebServices.  I can't comment on the "`hk2`" tag.  I can look into the --dependencies output to determine what part of the plugin needs this file.
      *  Imho, the problem is something to do with the naming as:  "ha-api.hk2**_-jar_**"
      * My guess is that that artefact was cached some place
      * Deleting it from the repository-server (Nexus) and deleting caches, eventually,  just means it come back later.
      *  That said, the generate seems to work OK and the WebServices are still working.
      *  it is very annoying.  For example Netbeans (v8.2 and v9.0-beta) keep flagging this error as a Gradle load problem.
      * A compile build works OK.  And the wsdl2java generate seems OK (_dunno_)

**Comment by aplatypus** [comment]
Here are the build file entries that track-back to this error.

Gradle build / plugins

```
plugins
{
	id 'no.nils.wsdl2java' version '0.10'
	//	see:	https://github.com/nilsmagnus/wsdl2java
	//			http://cxf.apache.org/docs/wsdl-to-java.html
}

apply plugin:	'no.nils.wsdl2java'
```

Gradle build / dependencies ...

```
dependencies {
           :
    compile    group: 'com.sun.xml.ws',    name: 'jaxws-tools',    version: jaxws_tools_version
    compile    group: 'com.sun.xml.ws',    name: 'jaxws-rt',         version: jaxws_rt_version
         :
}
```

Gradle dependencies output
*  **org.glassfish.ha:ha-api:3.1.9**

```
:
|
+--- com.sun.xml.ws:jaxws-tools:2.2.10
|    +--- com.sun.xml.ws:jaxws-rt:2.2.10
|    |    +--- javax.xml.bind:jaxb-api:2.2.12-b140109.1041
|    |    +--- javax.xml.ws:jaxws-api:2.2.11
|    |    |    +--- javax.xml.bind:jaxb-api:2.2.9 -> 2.2.12-b140109.1041
|    |    |    +--- javax.xml.soap:javax.xml.soap-api:1.3.5 -> 1.3.7
|    |    |    +--- javax.annotation:javax.annotation-api:1.2-b03 -> 1.2
|    |    |    \--- javax.jws:jsr181-api:1.0-MR1
|    |    +--- javax.xml.soap:javax.xml.soap-api:1.3.7
|    |    +--- javax.annotation:javax.annotation-api:1.2
|    |    +--- javax.jws:jsr181-api:1.0-MR1
|    |    +--- com.sun.xml.bind:jaxb-core:2.2.10-b140802.1033
|    |    +--- com.sun.xml.bind:jaxb-impl:2.2.10-b140802.1033
|    |    +--- com.sun.xml.ws:policy:2.4
|    |    +--- org.glassfish.gmbal:gmbal-api-only:3.1.0-b001
|    |    |    \--- org.glassfish.external:management-api:3.0.0-b012
|    |    +--- org.jvnet.staxex:stax-ex:1.7.7
|    |    |    +--- javax.xml.stream:stax-api:1.0-2
|    |    |    \--- javax.activation:activation:1.1 -> 1.1.1
|    |    +--- com.sun.xml.stream.buffer:streambuffer:1.5.3
|    |    |    \--- org.jvnet.staxex:stax-ex:1.7.1 -> 1.7.7 (*)
|    |    +--- org.jvnet.mimepull:mimepull:1.9.4
|    |    +--- com.sun.xml.fastinfoset:FastInfoset:1.2.13
|    |    +--- org.glassfish.ha:ha-api:3.1.9
|    |    +--- com.sun.xml.messaging.saaj:saaj-impl:1.3.25
:   :
```

**Comment by jjohannes** [comment]
Thanks @aplatypus for the additional information. It's very helpful!

First I think this is not a regression, but a specific dependency caching issue that exists for a long time already. It is triggered by a combination of two things (1) switching from one Gradle version to another and (2) specific behavior of you local Nexus (for which I don't know why it is behaving like it is)

When you switch Gradle versions, it can happen that Gradle needs to re-create its dependency metadata cache. Because we sometimes add more information to the cache. That's the case if you switch from `4.4` to `4.6`. Gradle is then looking in it's artfiact cache if it still has the original metadata and jar files. If it finds such a file, it checks the SHA1 to verify that it matches what's in the remote repository. If it can't get the SHA1, it downloads the dependency again. And only if it does not know the SHA1 yet, it will attempt to download a `.sha1` file.

Now a couple of things are happening:
1. The `ha-api` pom contains the entry `<packaging>hk2-jar</packaging>`. Because of that Gradle looks for a `ha-api-3.1.9.hk2-jar`. On maven central however, there is only `ha-api-3.1.9.jar`. Which is ok for Gradle, we just take `ha-api-3.1.9.jar` as the default. And if we ever get to the point where we request the `sha1` file (only if we switch versions and do not have the sha already, which is also unlikely for maven central) we would look for `ha-api-3.1.9.jar.sha1` (which exisits and is fine on maven central).
2. But, on your Nexus, the file `ha-api-3.1.9.hk2-jar` exists. So Gradle takes that and when it needs the sha1, after switching Gradle versions, it requests `ha-api-3.1.9.hk2-jar.sha1`.
3. For some reasons the `ha-api-3.1.9.hk2-jar.sha1` is broken, or the Nexus is not delivering it as Gradle would expect it. This might be something you could fix on your side. The reason why you did not see this problem before is that Gradle never requested the sha1 file before. The problem is possibly there for a long time already.
4. If Gradle gets some content for `ha-api-3.1.9.hk2-jar.sha1` it can't process, it fails the build with an error. I think this behavior is to harsh. Gradle could just log the error and go on downloading the dependency again (since we cannot find out if the SHA1 matches). This is what Gradle already does if the `.sha1` file does not exist at all. I have created a PR to change this behavior in Gradle 4.8.

Bottom line:
1. You might want to check your Nexus and fix the problem there if you can.
2. Workaround if you can not fix (1) for now: If you start using Gradle 4.6 or 4.7 with an empty cache (i.e. empty Gradle home) you won't run into the error
3. With Gradle 4.8 if you can not fix (1): The error will be ignored. Gradle won't be able to reuse some modules it already has in the local cache and instead re-download them. Which won't be an issue for you I assume as it uses your internal Nexus.
